database/pgx/pgx.go

```diff
@@ -56,8 +58,9 @@ type Config struct {
 }
 
 type Postgres struct {
-	// Locking and unlocking need to use the same connection
-	conn     *sql.Conn
+	conn *pgx.Conn
```

Comment:
Just realized that changing this a from a `*sql.Conn` to a `*pgx.Conn` will make it harder to [factor out common db driver logic](https://github.com/golang-migrate/migrate/issues/549) between the pgx and postgres drivers...

I don't see any direct pgx functionality being used in this PR. Are there any plans to do so?
Can we go back to using `*sql.Conn`? We can use [`Raw()`](https://golang.org/pkg/database/sql/#Conn.Raw) or [`Conn()`](https://pkg.go.dev/github.com/jackc/pgx/v4@v4.11.0/stdlib#Conn.Conn) to get the underlying `*pgx.Conn` if necessary.

Reply:
The benefit of using the pgx.Conn over sql.Conn is you can get a pgx.Conn from a sql.DB -- but you can't get a sql.DB from a pgx.Conn. This is the issue I was running into with a program of mine that uses pgx directly, for some speed benefits and some benefits around de/serialisation that can't be done through `database/sql` 

That said, `lib/pg`'s README states that it's in maintenance mode, and to use pgx instead, and I'm not aware of any other popular postgres go libraries.

Reply:
Can a `pgx.Conn` be converted to a `sql.Conn`? I didn't see a way in the [docs](https://pkg.go.dev/github.com/jackc/pgx/v4@v4.11.0/stdlib). In your project(s) can you get a `sql.Conn` even though you're using pgx?

> That said, `lib/pg`'s README states that it's in maintenance mode, and to use pgx instead, and I'm not aware of any other popular postgres go libraries.

`migrate` needs to continue supporting `lib/pq` since there are [still active users](https://github.com/lib/pq/issues/1010#issuecomment-846286396) even though it's in maintenance mode. There isn't much of a difference between the pgx and postgres drivers (diff the source code to see for yourself how trivial the differences are) and contributors shouldn't need to update both drivers when making changes.

Reply:
The only way I could find to get a sql.Conn/DB in my program would be to open a separate connection using sql.Open instead of just passing in the pgx.Conn my program already created to do it's database operations, which just doesn't feel that clean to me.

Reply:
Let's leave the pgx db driver as is for now until we have a usecase for using `pgx.Conn` directly. Hopefully, we'll have the postgres and pgx db drivers refactored by then... Thanks again for the PR!

Reply:
Am I understanding correctly that you mean that we'll leave the driver as it, and not merge this? If so, I'm fine with the MR getting closed, since I realised sometime during this process that the driver doesn't even *have* to be in-tree, and I'm fine with maintaining my own seperate package with this MR's contents extracted out.

Reply:
Yeah, let's leave the driver as is for now and not merge this PR.

Reply:
You can get an `*sql.DB` from a pgx connection by getting its `ConnConfig` with `Config()` and then passing that to [`stdlib.OpenDB()`](https://pkg.go.dev/github.com/jackc/pgx/v4@v4.13.0/stdlib#OpenDB). Would this provide a path forward for this PR? I know it doesn't seem as clean as being able to adapt `pgx.Conn` directly, but without it it just leaves that are using the pgx "native" interface to have to convert to use stdlib themselves

Reply:
Won't `stdlib.OpenDB()` create a new connection instead of using the existing one?

Reply:
Yes but if a user who's using `pgx.Conn` everywhere else wants to call this library then they have to do that anyway

Reply:
When a db driver is created with an existing connection, the user's [expectation](https://github.com/golang-migrate/migrate/issues/578) is for that connection to be used by `migrate` but managed by the user.

Reply:
Ok, I see your point there. 

I guess my qualm is just that, without this PR, pgx users that weren't otherwise using the stdlib compatibility layer need to import it and then open a connection with it themselves in ordert to pass it to `NewWithDatabaseInstance`. And if you're not using the `WithDatabaseIstance` entrypoint then it's still an awkward way in as afaict the URL-style connstrings aren't compatible due to `migrate` looking for a `pgx://` scheme instead of `postgres://`, which pgx itself doesn't do. If you're already bought in to pgx and try to pick up `migrate`, it's tricky to figure out how to get the puzzle pieces to slot together. I think pushing for a unified pgx/pq driver that uses the `sql.DB` interface will leave that usecase unserved.

Reply:
Ugh.  Just ran into this ourselves.  We're using pgx/v5 without the stdlib compat layer, and it seems like migrate doesn't really support using an existing pgx connection.

Not a deal breaker, but it would have been nice not having to duplicate a bunch of connection setup/handling code, just for the brief migrate connection. :wink: